### PR TITLE
Step 4: Simple Spaced Repetition

### PR DESCRIPTION
**Pull request by Overbryd**
* Added spaced repetition, wrongly answered cards 1 minute,
    correct cards 15 times the amount of correct answers
* Refactored `Card` to live outside of `Deck`
* Added `Answer` to deal with answers in a `Session`
* Changed `Session#pile` to sort by answers and spaced repetition rules
* Added `Session#next_card_at` to check when the next card is due
* Added a new deck `data/decks/preschool.csv`

screen shot 2015-09-23 at 12 03 34

### CODE REVIEW

**Comment by Overbryd** [comment]
Whoops. Hold back, just realised one problem:

I misread the description on Step 4, and I multiply the interval instead of doubling it every time.
Fix is on its way.

File: Gemfile [hunk 1]
```diff
@@ -6,5 +6,6 @@ group :test do
   gem "rack-test", require: "rack/test"
   gem "mocha", require: false
   gem "capybara", require: false
+  gem "timecop", require: false
```
Comment (plexus) [inline Gemfile]: good call, perfect use case for Timecop!

File: lib/session.rb [hunk 2]
```diff
@@ -38,19 +38,40 @@ def deck
...
 
+  def answers
+    @answers.map(&:freeze).freeze
+  end
+
   def pile
-    deck.cards.zip(answers).reject { |_, answered| answered }.map(&:first)
+    zipped = deck.cards.zip(@answers)
+    zipped.sort! do |(_, a), (_, b)|
+      a = a.nil? ? a.to_i : a.repeat_at.to_i
```
Comment (plexus) [inline lib/session.rb]: Why `a.to_i` if you know that `a` is `nil`? Wouldn't simply writing `0` be more clear?

File: lib/answer.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  attr_reader :interval, :at
+
+  def initialize(interval = 1, at = Time.now)
+    @interval, @at = interval, at
+  end
+
+  def correct
+    @interval = interval == 1 ? 15 : interval * 2
+    @at = Time.now
```
Comment (plexus) [inline lib/answer.rb]: Seems like a great candidate for being a Value Object.

File: lib/session.rb [hunk 4]
```diff
@@ -38,19 +38,40 @@ def deck
...
+  end
+
+  def next_card_at
+    return Time.now unless answer = answers[pile.first.id]
+    answer.repeat_at
   end
 
-  def answer(card_id)
+  def answer(card_id:, flag:)
     return if card_id.nil?
```
Comment (plexus) [inline lib/session.rb]: do you need this nil check? should you?

**Comment by plexus** [comment]
Code wise this looks very good. Design-wise I probably would have made `Answer` an (immutable) value object, and have used a null-object `Answer` for cards that don't have an answer yet, instead of using `nil`. That would have done away with some of the type/nil checks you now have. However I do think your approach is valid as well.

File: lib/session.rb (hunk 2 above)
Comment (Overbryd) [inline lib/session.rb]: True, of course much better.

File: lib/session.rb (hunk 4 above)
Comment (Overbryd) [inline lib/session.rb]: Since this method is supplied with user input, a nil value is possible. Therefore I also tested the nil case.

**Comment by Overbryd** [comment]
Great. Thank you for pushing the review further.

`Answer` might as well be an immutable value object, yet I found not enough evidence for refactoring towards that.
Personally I think that is definitely the next thing I would do, whenever I need more from `@answers` (e.g. statistics per session). But since `Session#pile` is the sole use case of `Answer` nil just does a great job.
I also think best would be if I had started with `Answer` being an immutable value object right away, so this refactoring wouldn't pose.
